### ISSUE TITLE
Feature Request: Lora compability converter

### ISSUE THREAD

**Issue by TheOnlyHolyMoly** [kohya-ss/sd-scripts#558]
Hello,

I have made an observation with lora files: Loras trained on 2.1 models will only work with 2.1 models and loras trained on 1.5 models will only work in combination with 1.5 models. Most likely I'm just stating the obvious but I wasnt aware of the fact, especially since a1111 does not validate if the selected lora is compatible with the selected model. It'll just throw cryptic errors instead.

So we are having a double incompability as not only loras will work with 1.5 OR 2.1 model, but also they cannot be combined with each other if they were trained on different model versions.

Anyhow, I believe there is so much lora content created upon 1.5 models that the majority of us will be bound to staying with 1.5 models due to the large pool of compatible loras. Since you know so much about training and creating the training tools, I was wondering if you can think of a lora wrapper/converters so 1.5 or 2.1 loras could be somehow made compatible with each other. 

Or maybe the training script could just load both 1.5. and 2.x model and output some kind of "joint lora"?

**Comment by TheOnlyHolyMoly** [comment]
I mean, the most complicated manual way of porting a lora from 1.5 to 2.x (and vice versa) would be to create a number of generated images from the 1.5 Lora and then retrain those images into a 2.x lora... but that would literally be a pain if done manually, also the outcome probably would not be as good as training from original data I guess... 
Looping in @KohakuBlueleaf and @vladmandic . Any thoughts on this?

**Comment by KohakuBlueleaf** [comment]
@TheOnlyHolyMoly 
TL;DR; 
Assume we have a lora trained for SD1.x series model. And we want to get a SD2.x version.
1. Retrain another lora for sd2.x is the easiest way
2. Train SD2.x lora from original data + SD1.x lora may produce better result (kind of distillation)
3. It is basically impossible to make SD2.x lora "from" SD1.x lora

for 3:
First I need to explain a thing: "Use 1 LoRA on different base Model doesn't make sense"
Since even a model trained with the same data and the same arch will have totally different parameters (if they have a different random seed when initialized).

So they will have a totally different 'position' in the 'space' of the 'model parameters.'
In this situation, a lora for A model will basically crash B model.

So, why we can use something like merge or use lora on different model? Since All of our model, are trained from SD1.x or SD2.x

It should be "Which means, for all SD1.x series models, they start from the same 'point' (SD1.0), and since the point is close enough to the 'best solution.' (or, the global minimum in the gradient descent) 
So all the model trained from it are very close to each other. Which means you can assume that "move them with same direction" (same lora model) will give quite same effect.

Back to SD1.x and SD2.x, apart from the differences between the arch, they are at the totally different "position" in the "space". And transform between 2 space are too hard to get (it is somekind of 860,000,000 dim space)
So transform lora/lyco model between 2 arch is basically impossible.

But yes, since I say "basically" I will not say it is "totally" impossible, but I will say it is not worth to do it. Train a new model on new arch are way more easy. And If you want to get familiar result, just use some distillation trick.

**Comment by TheOnlyHolyMoly** [comment]
Great explanation, so basically the point is the CD will not run in my minidisk player or vice versa. Understood.

My greatest concern was that people that originally cooked Loras might be moving on at some time instead of spending time on retraining their original images again for a 2.x lora (which I can perfectly understand). So I guess there would be some value in creating a more automated lora replication solution (generation + training based on generated images) solution? Intentionally saying more automated, not fully automated.

**Comment by vladmandic** [comment]
i agree that converting lora from/to sd15 and sd21 is not desirable as even in the best case, results would be suboptimal.
but we should present a clear message to user - @KohakuBlueleaf this is more for you...

right now if i try to load sd21 lora with sd15 model first we get 10+ screen dumps which should be better managed:

```diff
-             logger.warn(f'key failed to match: {key_diffusers}')
+             logger.debug(f'key failed to match: {key_diffusers}')
-         print(shared.sd_model.lyco_layer_mapping)
-         print(f"Failed to match keys when loading Lyco {filename}: {keys_failed_to_match}")
+         logger.error(f"Failed to match keys when loading Lyco {filename}: {len(keys_failed_to_match)}")
```

and finally it fails in `lyco_apply_weights` with:
> RuntimeError: The size of tensor a (768) must match the size of tensor b (1024) at non-singleton dimension 1

i suggest to add some sanity check in `lyco_apply_weights` and print error that user can understand

**Comment by KohakuBlueleaf** [comment]
> i agree that converting lora from/to sd15 and sd21 is not desirable as even in the best case, results would be suboptimal. but we should present a clear message to user - @KohakuBlueleaf this is more for you...
> 
> right now if i try to load sd21 lora with sd15 model first we get 10+ screen dumps which should be better managed:
> 
> ```diff
> -             logger.warn(f'key failed to match: {key_diffusers}')
> +             logger.debug(f'key failed to match: {key_diffusers}')
> -         print(shared.sd_model.lyco_layer_mapping)
> -         print(f"Failed to match keys when loading Lyco {filename}: {keys_failed_to_match}")
> +         logger.error(f"Failed to match keys when loading Lyco {filename}: {len(keys_failed_to_match)}")
> ```
> 
> and finally it fails in `lyco_apply_weights` with:
> 
> > RuntimeError: The size of tensor a (768) must match the size of tensor b (1024) at non-singleton dimension 1
> 
> i suggest to add some sanity check in `lyco_apply_weights` and print error that user can understand

Ok
BTW As I know, built-in lora also has same kind of error?

**Comment by bmaltais** [comment]
Making LoRA for SD2.1 is very difficult for some strange reasons. Results are very good in 1.5 but are garbage unser 2.1... this is why people avoid training for 2.1 because it does not produce good results... probably due to the censoring of all the naughty dataset bits.

**Comment by TheOnlyHolyMoly** [comment]
@derspanier thought you might find this interesting

**Comment by bmaltais** [comment]
I tried again with the latest code from kohya and even with the latest LyCORIS network LoHA, LoKr and iA3 and no luck. Trained LoRA still produce garbage results with SD2.1... I am giving up ;-)

**Comment by KohakuBlueleaf** [comment]
> Making LoRA for SD2.1 is very difficult for some strange reasons. Results of trained LoRA are very good in 1.5 but are mostly garbage unser 2.1... this is why people avoid training for 2.1 because it does not produce good results... probably due to the censoring of all the naughty dataset bits.

I think the problem is the CLIP used in SD2.1 is too bad.
(Maybe we need Datacomp one)

**Comment by vladmandic** [comment]
@KohakuBlueleaf you think its the clip? not sure if datacomp models would slide in neatly as-is or would need some adjustments, but either way it should be doable.

**Comment by KohakuBlueleaf** [comment]
@vladmandic definitely need adjustments
and datacomp only have ViT-B and L one
SDv2 use CLIP-H one (and trained by Laion 2b which is bad)

I think the CLIP directly trained on Laion2B is quite bad

**Comment by vladmandic** [comment]
its been a while since i last checked their clips, but it seems they have more variations released nowadays: <https://github.com/mlfoundations/open_clip/blob/main/docs/datacomp_models.md>

**Comment by KohakuBlueleaf** [comment]
> its been a while since i last checked their clips, but it seems they have more variations released nowadays: https://github.com/mlfoundations/open_clip/blob/main/docs/datacomp_models.md

Still only ViT-L one and ViT-B one
the scale is for "different dataset scale" not model scale.
since datacomp is for dataset researching
